src/GafferSceneUI/AttributeQueryUI.cpp

```diff
@@ -34,74 +34,235 @@
+		items.push_back( MenuItem( "Array/Float", createVectorDataPlug( name, direction, new IECore::FloatVectorData(), flags ) ) );
+		items.push_back( MenuItem( "Array/Int", createVectorDataPlug( name, direction, new IECore::IntVectorData(), flags ) ) );
+		items.push_back( MenuItem( "Array/" ) );
+		items.push_back( MenuItem( "Array/String", createVectorDataPlug( name, direction, new IECore::StringVectorData(), flags ) ) );
 	}
 
-	struct Register
+	return items;
+}
+
+struct Registration
+{
+	Registration()
 	{
-		Register()
+		GafferUI::NoduleLayout::registerCustomGadget( "GafferSceneUI.AttributeQueryUI.PlugAdder", & Registration::create );
+	}
+
+	static GafferUI::GadgetPtr create( Gaffer::GraphComponentPtr parent )
+	{
+		const GafferScene::AttributeQueryPtr query = IECore::runTimeCast< GafferScene::AttributeQuery >( parent );
+
+		if( ! query )
 		{
-			GafferUI::NoduleLayout::registerCustomGadget( "GafferUI.AttributeQueryUI.PlugAdder", & Register::create );
+			throw IECore::Exception( "AttributeQueryUI.PlugAdder requires an AttributeQuery" );
 		}
 
-		static GafferUI::GadgetPtr create( Gaffer::GraphComponentPtr parent )
-		{
-			const GafferScene::AttributeQueryPtr query = IECore::runTimeCast< GafferScene::AttributeQuery >( parent );
+		const GafferUI::GadgetPtr gadget( new GafferSceneUI::AttributeQueryUI::PlugAdder( *query ) );
+		return gadget;
+	}
+};
 
-			if( ! query )
-			{
-				throw IECore::Exception( "AttributeQueryPlugAdder requires an AttributeQuery" );
-			}
+Registration m_registration;
 
-			const GafferUI::GadgetPtr gadget( new AttributeQueryPlugAdder( *query ) );
-			return gadget;
-		}
-	};
+} // namespace
+
+namespace GafferSceneUI
+{
 
-	static Register m_register;
+const std::string& AttributeQueryUI::setupMenuTitle()
+{
+	static const std::string title( "Plug type" );
 
-	GafferScene::AttributeQueryPtr m_query;
-};
+	return title;
+}
+
+const std::vector< std::string >& AttributeQueryUI::setupMenuNames()
+{
+	static std::vector< std::string > names;
+
+	if( names.empty() )
+	{
+		const MenuItemContainer::nth_index< 0 >::type& items = menuItemContainer().get< 0 >();
+		std::transform( items.begin(), items.end(), std::back_inserter( names ), std::mem_fn( & MenuItem::getName ) );
+	}
+
+	return names;
+}
+
+bool AttributeQueryUI::setupFromMenuName( GafferScene::AttributeQuery& query, const std::string& name )
+{
+	if( name.empty() )
+	{
+		return false;
+	}
+
+	if( query.isSetup() )
+	{
+		return false;
+	}
+
+	const MenuItemContainer::nth_index< 1 >::type& items = menuItemContainer().get< 1 >();
+	const MenuItemContainer::nth_index< 1 >::type::iterator it = items.find( name );
 
-bool AttributeQueryPlugAdder::canCreateConnection( const Gaffer::Plug* const plug ) const
+	if( it == items.end() )
+	{
+		return false;
+	}
+
+	const Gaffer::ValuePlug* const plug = ( *it ).getPlug();
+
+	query.setup( plug );
```

Comment:
This needs to be done in an `UndoScope`, so that the operation can be undone with `Ctrl+U` in the UI :

```
UndoScope undoScope( query.ancestor<ScriptNode>() );
query.setup( plug );
```